modules/samples/watertank/component/api/src/main/java/org/homeautomation/watertank/api/DeviceTypeServiceImpl.java

```diff
@@ -0,0 +1,434 @@
+        String sensorTableName;
+        switch (sensorName) {
+            case DeviceTypeConstants.STREAM_RELAY:
+                sensorTableName = DeviceTypeConstants.RELAY_EVENT_TABLE;
+                break;
+            case DeviceTypeConstants.STREAM_WATERLEVEL:
+                sensorTableName = DeviceTypeConstants.WATERLEVEL_EVENT_TABLE;
+                break;
+            default:
+                return Response.status(Response.Status.BAD_REQUEST).entity("Invalid event stream").build();
+        }
+
+        try {
+            if (!APIUtil.getDeviceAccessAuthorizationService().isUserAuthorized(new DeviceIdentifier(deviceId,
+                                                                                                     DeviceTypeConstants.DEVICE_TYPE))) {
+                return Response.status(Response.Status.UNAUTHORIZED.getStatusCode()).build();
+            }
+            List<SortByField> sortByFields = new ArrayList<>();
+            SortByField sortByField = new SortByField("time", SORT.ASC, false);
+            sortByFields.add(sortByField);
+            List<SensorRecord> sensorRecords = APIUtil.getAllEventsForDevice(sensorTableName, query, sortByFields);
+            return Response.status(Response.Status.OK.getStatusCode()).entity(sensorRecords).build();
+        } catch (AnalyticsException e) {
+            String errorMsg = "Error on retrieving stats on table " + sensorTableName + " with query " + query;
+            log.error(errorMsg);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()).entity(errorMsg).build();
+        } catch (DeviceAccessAuthorizationException e) {
+            log.error(e.getErrorMessage(), e);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Remove device type instance using device id
+     *
+     * @param deviceId unique identifier for given device type instance
+     */
+    @Path("/device/{deviceId}")
+    @DELETE
+    public Response removeDevice(@PathParam("deviceId") String deviceId) {
+        try {
+            DeviceIdentifier deviceIdentifier = new DeviceIdentifier();
+            deviceIdentifier.setId(deviceId);
+            deviceIdentifier.setType(DeviceTypeConstants.DEVICE_TYPE);
+            if (!APIUtil.getDeviceAccessAuthorizationService().isUserAuthorized(deviceIdentifier)) {
+                return Response.status(Response.Status.UNAUTHORIZED.getStatusCode()).build();
+            }
+            boolean removed = APIUtil.getDeviceManagementService().disenrollDevice(
+                    deviceIdentifier);
+            if (removed) {
+                return Response.ok().build();
+            } else {
+                return Response.status(Response.Status.NOT_ACCEPTABLE.getStatusCode()).build();
+            }
+        } catch (DeviceManagementException e) {
+            log.error(e.getErrorMessage(), e);
```

Comment:
repeating the same error message (i.e. e.getErrorMessage()) here, which can be found in the stack trace as well, is not a good practice.